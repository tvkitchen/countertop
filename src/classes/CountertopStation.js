import { v4 as uuid } from 'uuid'
import { countertopStates } from '../constants'
import { consoleLogger } from '../tools/loggers'
import { CountertopWorker } from './CountertopWorker'

class CountertopStation {
	logger = null

	kafka = null

	id = null

	Appliance = null

	applianceSettings = null

	workers = []

	state = ''

	/**
	 * Create a CountertopStation
	 *
	 * @param  {Class}  Appliance         The IAppliance class that this station will manage.
	 * @param  {Object} applianceSettings Settings for the appliance.
	 * @param  {Logger} options.logger    A logger with methods for all TV Kitchen logLevels.
	 */
	constructor(
		Appliance,
		applianceSettings = {},
		{
			logger = consoleLogger,
			kafka,
		} = {},
	) {
		if (Appliance === undefined) {
			throw new Error('CountertopStation requires an Appliance.')
		}
		this.Appliance = Appliance
		this.applianceSettings = applianceSettings
		this.logger = logger
		this.kafka = kafka
		this.id = `${Appliance.name}::${uuid()}`
		this.setState(countertopStates.STOPPED)
	}

	/**
	 * Replace the station's workers in order to fit a specified topology.
	 *
	 * @param  {CountertopToplogy} topology The topology to use.
	 * @return {CountertopWorker[]}         The new set of workers.
	 */
	invokeTopology = (topology) => {
		this.logger.trace(`CountertopStation<${this.id}>: invokeTopology()`)
		if (this.getState() !== countertopStates.STOPPED) {
			throw new Error('CountertopStations must be stopped before invoking a topology.')
		}
		const stationStreams = topology.streams.filter((stream) => stream.getMouth() === this)
		this.workers = stationStreams.map((stream) => new CountertopWorker(
			this.Appliance,
			this.applianceSettings,
			{
				stream,
				logger: this.logger,
				kafka: this.kafka,
			},
		))
		return this.workers
	}

	/**
	 * Start the station and begin data processing.
	 *
	 * @return {Boolean} Whether the station successfully started.
	 */
	start = async () => {
		this.logger.trace(`CountertopStation<${this.id}>: start()`)
		this.state = countertopStates.STARTING
		const promises = this.workers.map(async (worker) => worker.start())
		const started = (await Promise.all(promises)).every((result) => result)
		this.setState(started ? countertopStates.STARTED : countertopStates.ERRORED)
		return this.getState() === countertopStates.STARTED
	}

	/**
	 * Stop the countertop and halt all data processing.
	 *
	 * @return {Boolean} Whether the station successfully stopped.
	 */
	stop = async () => {
		this.logger.trace(`CountertopStation<${this.id}>: stop()`)
		this.setState(countertopStates.STOPPING)
		const promises = this.workers.map(async (worker) => worker.stop())
		const stopped = (await Promise.all(promises)).every((result) => result)
		this.setState(stopped ? countertopStates.STOPPED : countertopStates.ERRORED)
		return this.getState() === countertopStates.STOPPED
	}

	/**
	 * Get the data types that this station processes.
	 *
	 * This is determined by the Appliance that the station hosts.
	 *
	 * @return {String[]} The data types that this station processes.
	 */
	getInputTypes = () => this.Appliance.getInputTypes(this.applianceSettings)

	/**
	 * Get the data types that this station produces.
	 *
	 * This is determined by the Appliance that the station hosts.
	 *
	 * @return {String[]} The data types that this station produces.
	 */
	getOutputTypes = () => this.Appliance.getOutputTypes(this.applianceSettings)

	/**
	 * Get the current state of the CountertopStation.
	 *
	 * @return {String} The state of the CountertopStation.
	 */
	getState = () => this.state

	/**
	 * Set the current state of the CountertopStation.
	 *
	 * This is an internal method.
	 */
	setState = (state) => { this.state = state }

	/**
	 * Registers a listener to the CountertopStation for a given event type.
	 *
	 * Event types are defined in @tvkitchen/base-constants
	 *
	 * @param  {String} eventType  The type of event being listened to.
	 * @param  {Function} listener The listener to be registered for that event.
	 * @return {CountertopStation} The CountertopStation instance (to enable chaining).
	 */
	on = (eventType, listener) => {
		this.workers.forEach((worker) => worker.on(eventType, listener))
		return this
	}
}

export default CountertopStation
