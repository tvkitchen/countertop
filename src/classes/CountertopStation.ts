import { v4 as uuid } from 'uuid'
import { Kafka } from 'kafkajs'
import { CountertopStationState } from '../types'
import { CountertopStationStateError } from '../errors'
import { CountertopWorker } from './CountertopWorker'
import type { Logger } from '../types'
import type {
	ImplementedApplianceClass,
	ApplianceSettings,
} from './AbstractAppliance'
import type { CountertopTopology } from './CountertopTopology'

interface CountertopStationSettings {
	logger?: Logger;
	kafka: Kafka;
}

export class CountertopStation {
	public readonly settings: CountertopStationSettings

	public readonly id: string

	public readonly ApplianceClass: ImplementedApplianceClass

	public readonly applianceSettings?: ApplianceSettings

	#workers: CountertopWorker[]

	#state: CountertopStationState

	constructor(
		ApplianceClass: ImplementedApplianceClass,
		applianceSettings?: ApplianceSettings,
		settings: CountertopStationSettings = {
			kafka: new Kafka({ brokers: [] }),
		},
	) {
		this.ApplianceClass = ApplianceClass
		this.applianceSettings = applianceSettings
		this.settings = settings
		this.id = `${ApplianceClass.name}::${uuid()}`
		this.#state = CountertopStationState.Stopped
		this.#workers = []
	}

	/**
	 * Replace the station's workers in order to fit a specified topology.
	 *
	 * @param  {CountertopToplogy} topology The topology to use.
	 * @return {CountertopWorker[]}         The new set of workers.
	 */
	public invokeTopology(
		topology: CountertopTopology,
	): CountertopWorker[] {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: invokeTopology()`)
		if (this.state !== CountertopStationState.Stopped) {
			throw new CountertopStationStateError('CountertopStations must be stopped when invoking a topology.')
		}

		const stationStreams = topology.streams.filter(
			(stream) => stream.mouth === this,
		)
		this.#workers = stationStreams.map((stream) => (
			new CountertopWorker(
				this.ApplianceClass,
				this.applianceSettings,
				{
					stream,
					logger: this.settings.logger,
					kafka: this.settings.kafka,
				},
			)
		))
		return this.#workers
	}

	/**
	 * Start the CountertopStation and begin data processing.
	 */
	public async start(): Promise<void> {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: start()`)
		this.#state = CountertopStationState.Starting
		try {
			await this.startAllWorkers()
		} catch (e: unknown) {
			this.#state = CountertopStationState.Errored
			await this.stopAllWorkers()
			throw e
		}
		this.#state = CountertopStationState.Started
	}

	/**
	 * Stop the CountertopStation and halt all data processing.
	 */
	public async stop(): Promise<void> {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: stop()`)
		this.#state = CountertopStationState.Stopping
		await this.stopAllWorkers()
	}

	/**
	 * Get the data types that this station processes.
	 *
	 * This is determined by the Appliance that the station hosts.
	 */
	public getInputTypes() {
		return this.ApplianceClass.getInputTypes(this.applianceSettings)
	}

	/**
	 * Get the data types that this station produces.
	 *
	 * This is determined by the Appliance that the station hosts.
	 */
	public getOutputTypes() {
		return this.ApplianceClass.getOutputTypes(this.applianceSettings)
	}

	public get state() {
		return this.#state
	}

	/**
	 * Registers a listener to the CountertopStation for a given event type.
	 *
	 * Event types are defined in @tvkitchen/base-constants
	 */
	public on(
		eventType: string,
		listener: (...args: unknown[]) => void,
	): this {
		this.#workers.forEach(
			(worker) => worker.on(eventType, listener),
		)
		return this
	}

	private async startAllWorkers(): Promise<void> {
		await Promise.all(
			this.#workers.map(
				async (worker) => worker.start(),
			),
		)
	}

	private async stopAllWorkers(): Promise<void> {
		await Promise.all(
			this.#workers.map(
				async (worker) => worker.stop(),
			),
		)
	}
}
