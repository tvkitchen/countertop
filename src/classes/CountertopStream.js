import assert from 'assert'
import { v4 as uuid } from 'uuid'
import {
	getLongestStreamLength,
	getSourcesFromStreamMap,
} from '../tools/utils/countertop'

class CountertopStream {
	id = null

	tributaries = null

	mouth = null

	source = null

	/**
	 * Create a new CountertopStream.
	 *
	 * CountertopStreams are made up of a distinct stream of stations (e.g. Appliance instances).
	 * This allows a given countertop to have multiple sources of video data being processed
	 * simultaneously.
	 *
	 * The tributary map does not need to be complete -- there can be inputs with no associated
	 * tributary streams.
	 *
	 * The tributary map cannot contain irrelevant streams (the output map should match the inputs).
	 *
	 * @param  {CountertopStation} station     The station at the end of the stream.
	 * @param  {Map}               tributaries The streams that are flowing into the final station.
	 */
	constructor(station, tributaries = new Map()) {
		Array.from(tributaries.keys()).forEach(
			(outputType) => assert(
				station.getInputTypes().includes(outputType),
				'The tributaries contained an invalid stream type.',
			),
		)
		this.id = `CountertopStream::${uuid()}`
		this.mouth = station
		this.tributaries = tributaries
		if (station.getInputTypes().length === 0) {
			this.source = station
		} else {
			this.source = getSourcesFromStreamMap(tributaries).pop() ?? null
		}
	}

	/**
	 * Get the origin / first station in the stream.
	 *
	 * @return {CountertopStation} The origin station in the stream.
	 */
	getSource = () => this.source

	/**
	 * Get the last station in the stream.
	 * This is the station that will produce new data, and determines the stream's
	 * output types.
	 *
	 * @return {CountertopStation} The last station in the stream.
	 */
	getMouth = () => this.mouth

	/**
	 * Get the Map of tributary streams that feed into the mouth station.
	 *
	 * @return {Map} The map of streams keyed by the input types they feed into.
	 */
	getTributaryMap = () => this.tributaries

	/**
	 * Get an array of tributaries that feed into the mouth station.
	 *
	 * @return {CountertopStream[]} The array of tributary streams.
	 */
	getTributaryArray = () => [...this.tributaries.values()]

	/**
	 * Get the maximum length of the longest stream path between the stream's source and mouth.
	 *
	 * @return {Number} The length of the stream
	 */
	getLength = () => getLongestStreamLength(this.getTributaryArray()) + 1

	/**
	 * Checks if a given station exists at any point within the stream.
	 *
	 * @param  {CountertopStation} station The station being searched for.
	 * @return {Boolean}                   The result of the search.
	 */
	includesStation = (station) => this.mouth === station
		|| this.getTributaryArray().some((tributary) => tributary.includesStation(station))

	/**
	 * Get the output types produced by this stream.
	 *
	 * @return {String[]} The output types.
	 */
	getOutputTypes = () => this.mouth.getOutputTypes()
}

export default CountertopStream
