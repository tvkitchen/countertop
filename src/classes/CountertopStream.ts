import assert from 'assert'
import { v4 as uuid } from 'uuid'
import {
	getLongestStreamLength,
	getSourcesFromStreams,
} from '../tools/utils/countertop'

export class CountertopStream {
	public readonly id: string

	public readonly tributaries: Map<string, CountertopStream>

	public readonly mouth: any // this will be a CountertopStation once defined

	public readonly source: any // this will be a CountertopStation once defined

	/**
	 * Create a new CountertopStream.
	 *
	 * CountertopStreams are made up of a distinct stream of stations (e.g. Appliance instances).
	 * This allows a given Countertop to have multiple sources of video data being processed
	 * simultaneously.
	 *
	 * The tributary map does not need to be complete -- there can be inputs with no associated
	 * tributary streams.
	 *
	 * The tributary map cannot contain irrelevant streams (the output map should match the inputs).
	 *
	 * @param  {CountertopStation} station The station at the end of the stream.
	 * @param  {Map<string, CountertopStream>} tributaries The streams providing data for each input.
	 */
	constructor(
		countertopStation: any, // this will be a CountertopStation once defined
		tributaries = new Map<string, CountertopStream>(),
	) {
		Array.from(tributaries.keys()).forEach(
			(outputType) => assert(
				// TODO: remove this when CountertopStation has type definitions
				// eslint-disable-next-line
				countertopStation.getInputTypes().includes(outputType),
				'The tributaries contained an invalid stream type.',
			),
		)
		this.id = `CountertopStream::${uuid()}`
		// TODO: remove this when CountertopStation has type definitions
		// eslint-disable-next-line
		this.mouth = countertopStation
		this.tributaries = tributaries
		// TODO: remove this when CountertopStation has type definitions
		// eslint-disable-next-line
		if (countertopStation.getInputTypes().length === 0) {
			// TODO: remove this when CountertopStation has type definitions
			// eslint-disable-next-line
			this.source = countertopStation
		} else {
			// TODO: remove this when CountertopStation has type definitions
			// eslint-disable-next-line
			this.source = getSourcesFromStreams(this.getTributaryArray()).pop() ?? null
		}
	}

	/**
	 * Get an array of tributaries that feed into the mouth station.
	 *
	 * @return {CountertopStream[]} The array of tributary streams.
	 */
	public getTributaryArray(): CountertopStream[] {
		return [...this.tributaries.values()]
	}

	/**
	 * Get the maximum length of the longest stream path between the stream's source and mouth.
	 *
	 * @return {Number} The length of the stream
	 */
	public getLength() {
		return getLongestStreamLength(this.getTributaryArray()) + 1
	}

	/**
	 * Checks if a given station exists at any point within the stream.
	 *
	 * @param  {CountertopStation} station The station being searched for.
	 * @return {Boolean}                   The result of the search.
	 */
	public includesStation(station: any): boolean {
		return this.mouth === station
			|| this.getTributaryArray().some(
				// TODO: remove this when CountertopStation has type definitions
				// eslint-disable-next-line
				(tributary: any) => tributary.includesStation(station),
			)
	}

	/**
	 * Checks if a given sub-stream exists at any point within the stream.
	 */
	public includesStream(subStream: CountertopStream): boolean {
		return this.getTributaryArray().includes(subStream)
			|| this.getTributaryArray().some(
				(tributary: CountertopStream): boolean => tributary.includesStream(subStream),
			)
	}

	/**
	 * Get the output types produced by this stream.
	 *
	 * @return {String[]} The output types.
	 */
	public getOutputTypes(): any {
		// TODO: remove this when CountertopStation has type definitions
		// eslint-disable-next-line
		return this.mouth.getOutputTypes()
	}
}
