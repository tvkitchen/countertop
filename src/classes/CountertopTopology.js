import CountertopStream from './CountertopStream'
import {
	getCollectiveOutputTypes,
	getStationsThatConsumeTypes,
	getLongestStreamLength,
	filterStreamsContainingStation,
	generateTributaryMaps,
	getStreamOutputMap,
	getSourceStations,
	identifyObsoleteStreams,
	pruneStreams,
} from '../tools/utils/countertop'

/**
 * CountertopToplogies are a complete set of valid *paths* that data might take across a given
 * array of CountertopStations.
 *
 * The distinction of paths (as opposed to edges / direct steps) is important, as a given payload
 * will have a source, and only payloads with a common source should be considered together in
 * the case of dual-input appliances.
 *
 * In many data processing pipelines, topologies need to be manually specified by the developer
 * configuring the system. In this case, the topologies are automatically generated based on which
 * appliances produce outputs that are valid inputs to others.
 *
 * The logic for generating a CountertopTopology starts in the static `generateStreams` method.
 */
class CountertopTopology {
	stations = []

	streams = []

	/**
	 * Create a CountertopTopology.
	 *
	 * @param  {CountertopStation[]} stations The stations used to define the topology.
	 */
	constructor(stations = []) {
		this.stations = stations
		this.streams = CountertopTopology.generateStreams(this.stations)
	}

	/**
	 * Extend a set of streams by a single station.
	 *
	 * This will only create streams that pull from the same source.
	 * This will generate and return any valid streams that extend a given stream set.
	 * This will ONLY return new streams, which means if no valid streams extensions exist, it will
	 * return an empty array.
	 * This will not create streams that re-visit a given station (no loops).
	 *
	 * @param  {CountertopStream[]}  streams The set of streams to be extended.
	 * @param  {CountertopStation[]} station The station being used to extend.
	 * @return {CountertopStream[]}          The extended streams.
	 */
	static extendStreamsByStation = (streams, station) => {
		const streamOutputMap = getStreamOutputMap(
			filterStreamsContainingStation(streams),
		)
		const tributaryMaps = generateTributaryMaps(
			station,
			streamOutputMap,
		)
		return tributaryMaps.map(
			(tributaryMap) => new CountertopStream(station, tributaryMap),
		)
	}

	/**
	 * Fully extends a set of streams to flow through a set of new stations.
	 *
	 * This will only create streams that pull from the same source.
	 * This will not create streams that re-visit a given station (no loops).
	 * This will create distinct streams for incremental steps from a source (e.g. A->B, and A->B->C).
	 *
	 * @param  {CountertopStream[]}  streams  The base set of streams being extended.
	 * @param  {CountertopStation[]} stations The stations being used to extend the streams.
	 * @return {CountertopStream[]}           The complete set of streams.
	 */
	static generateStreams = (stations, streams = []) => {
		let nextStreams = []
		if (streams.length === 0) {
			nextStreams = CountertopTopology.generateSourceStreams(stations)
		} else {
			const extentionLength = 1 + getLongestStreamLength(streams)
			const outputTypes = getCollectiveOutputTypes(streams)
			const nextStations = getStationsThatConsumeTypes(stations, outputTypes)
			nextStreams = nextStations.flatMap(
				(station) => CountertopTopology.extendStreamsByStation(streams, station),
			)
				// Remove any new streams that aren't long enough
				.filter((stream) => stream.getLength() === extentionLength)
		}
		// Prune any past streams whose tributaries are explicit subsets of new streams
		const obsoleteStreams = identifyObsoleteStreams(streams, nextStreams)
		const prunedStreams = pruneStreams(streams, obsoleteStreams)
		const prunedNextStreams = pruneStreams(nextStreams, obsoleteStreams)

		// If there were any new streams, iterate again
		if (prunedNextStreams.length !== 0) {
			return CountertopTopology.generateStreams(
				stations,
				prunedStreams.concat(prunedNextStreams),
			)
		}
		return prunedStreams
	}

	/**
	 * Generates streams that only include source stations.
	 *
	 * These streams will, by defintiion, be of length 1.
	 *
	 * @param  {CountertopStations[]} stations The stations to be processed.
	 * @return {CountertopStreams[]}           The resulting streams.
	 */
	static generateSourceStreams = (stations) => getSourceStations(stations)
		.map((station) => new CountertopStream(station))
}

export default CountertopTopology
