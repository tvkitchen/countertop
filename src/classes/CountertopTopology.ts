import {
	getCollectiveOutputTypes,
	getStationsThatConsumeTypes,
	getLongestStreamLength,
	filterStreamsContainingStation,
	generateTributaryMaps,
	getStreamOutputMap,
	getSourceStations,
	identifyObsoleteStreams,
	pruneStreams,
} from '../tools/utils/countertop'
import { CountertopStream } from './CountertopStream'
import { CountertopStation } from './CountertopStation'

/**
 * CountertopToplogies are a complete set of valid *paths* that data might take across a given
 * array of CountertopStations.
 *
 * The distinction of paths (as opposed to edges / direct steps) is important, as a given payload
 * will have a source, and only payloads with a common source should be considered together in
 * the case of dual-input appliances.
 *
 * In many data processing pipelines, topologies need to be manually specified by the developer
 * configuring the system. In this case, the topologies are automatically generated based on which
 * appliances produce outputs that are valid inputs to others.
 *
 * The logic for generating a CountertopTopology starts in the static `generateStreams` method.
 */
export class CountertopTopology {
	stations: CountertopStation[]

	streams: CountertopStream[]

	constructor(stations: CountertopStation[] = []) {
		this.stations = stations
		this.streams = CountertopTopology.generateStreams(this.stations)
	}

	/**
	 * Fully extends a set of streams to flow through a set of new stations.
	 *
	 * This will only create streams that pull from the same source.
	 * This will not create streams that re-visit a given station (no loops).
	 * This will create distinct streams for incremental steps from a source (e.g. A->B, and A->B->C).
	 *
	 * @param  {CountertopStream[]}  streams  The base set of streams being extended.
	 * @param  {CountertopStation[]} stations The stations being used to extend the streams.
	 * @return {CountertopStream[]}           The complete set of streams.
	 */
	public static generateStreams(
		stations: CountertopStation[],
		streams: CountertopStream[] = [],
	): CountertopStream[] {
		let nextStreams = []
		if (streams.length === 0) {
			nextStreams = CountertopTopology.generateSourceStreams(stations)
		} else {
			const extentionLength = 1 + getLongestStreamLength(streams)
			const outputTypes = getCollectiveOutputTypes(streams)
			const nextStations = getStationsThatConsumeTypes(stations, outputTypes)
			nextStreams = nextStations.flatMap(
				// Disable until countertop utilities get types
				// eslint-disable-next-line
				(station) => CountertopTopology.extendStreamsByStation(streams, station),
			)
				// Remove any new streams that aren't long enough
				.filter((stream) => stream.getLength() === extentionLength)
		}
		// Prune any past streams whose tributaries are explicit subsets of new streams
		const obsoleteStreams = identifyObsoleteStreams(streams, nextStreams)
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		const prunedStreams = pruneStreams(streams, obsoleteStreams)
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		const prunedNextStreams = pruneStreams(nextStreams, obsoleteStreams)

		// If there were any new streams, iterate again
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		if (prunedNextStreams.length !== 0) {
			return CountertopTopology.generateStreams(
				stations,
				// Disable until countertop utilities get types
				// eslint-disable-next-line
				prunedStreams.concat(prunedNextStreams),
			)
		}
		// Disable until countertop utilities get types
		// eslint-disable-next-line
		return prunedStreams
	}

	/**
	 * Extend a set of streams by a single station.
	 *
	 * This will only create streams that pull from the same source.
	 * This will generate and return any valid streams that extend a given stream set.
	 * This will ONLY return new streams, which means if no valid streams extensions exist, it will
	 * return an empty array.
	 * This will not create streams that re-visit a given station (no loops).
	 *
	 * @param  {CountertopStream[]}  streams The set of streams to be extended.
	 * @param  {CountertopStation[]} station The station being used to extend.
	 * @return {CountertopStream[]}          The extended streams.
	 */
	private static extendStreamsByStation(
		streams: CountertopStream[],
		station: CountertopStation,
	): CountertopStream[] {
		const streamOutputMap = getStreamOutputMap(
			filterStreamsContainingStation(streams, station),
		)
		const tributaryMaps = generateTributaryMaps(
			station,
			streamOutputMap,
		)
		return tributaryMaps.map(
			// Disable until countertop utilities get types
			// eslint-disable-next-line
			(tributaryMap) => new CountertopStream(station, tributaryMap),
		)
	}

	/**
	 * Generates streams that only include source stations.
	 *
	 * These streams will, by defintiion, be of length 1.
	 *
	 * @param  {CountertopStations[]} stations The stations to be processed.
	 * @return {CountertopStreams[]}           The resulting streams.
	 */
	private static generateSourceStreams(stations: CountertopStation[]) {
		return getSourceStations(stations)
			// Disable until countertop utilities get types
			// eslint-disable-next-line
			.map((station) => new CountertopStream(station))
	}
}
