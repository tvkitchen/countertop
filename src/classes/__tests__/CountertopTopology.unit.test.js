import CountertopTopology from '../CountertopTopology'
import { CountertopStation } from '../CountertopStation'
import { CountertopStream } from '../CountertopStream'

import {
	generateMockAppliance,
	normalizeStreams,
} from '../../tools/utils/jest'

describe('CountertopTopology #unit', () => {
	describe('constructor', () => {
		it('Should generate no streams if provided no stations', () => {
			const topology = new CountertopTopology()
			expect(topology.streams).toEqual([])
		})
	})

	describe('extendStreamsByStation', () => {
		it('Should extend streams correctly', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const streamA = new CountertopStream(stationA)
			const streams = CountertopTopology.extendStreamsByStation([streamA], stationB)
			expect(streams.length).toBe(1)
		})
		it('Should extend streams even if a station does not have inputs', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['baz'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const streamA = new CountertopStream(stationA)
			const streams = CountertopTopology.extendStreamsByStation([streamA], stationB)
			expect(streams.length).toBe(1)
		})
	})

	describe('generateStreams', () => {
		it('Should generate the correct streams', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stations = [stationA, stationB]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
		it('Should not generate streams for stations with no inputs', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['baz'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stations = [stationA, stationB]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
		it('Should not generate redundant partial streams for stations with multiple inputs', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const ApplianceC = generateMockAppliance({
				inputTypes: ['foo', 'bar'],
				outputTypes: ['baz'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stationC = new CountertopStation(ApplianceC)
			const stations = [stationA, stationB, stationC]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
		it('Should maintain complete toplogies after redundant partial streams are removed', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			})
			const ApplianceC = generateMockAppliance({
				inputTypes: ['foo', 'bar'],
				outputTypes: ['baz'],
			})
			const ApplianceD = generateMockAppliance({
				inputTypes: ['baz'],
				outputTypes: ['bop'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stationC = new CountertopStation(ApplianceC)
			const stationD = new CountertopStation(ApplianceD)
			const stations = [stationA, stationB, stationC, stationD]
			const streams = CountertopTopology.generateStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
	})

	describe('generateSourceStreams', () => {
		it('Should generate streams for any source stations', () => {
			const ApplianceA = generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			})
			const ApplianceB = generateMockAppliance({
				inputTypes: ['baz'],
				outputTypes: ['bar'],
			})
			const stationA = new CountertopStation(ApplianceA)
			const stationB = new CountertopStation(ApplianceB)
			const stations = [stationA, stationB]
			const streams = CountertopTopology.generateSourceStreams(stations)
			expect(normalizeStreams(streams, stations)).toMatchSnapshot()
		})
	})
})
