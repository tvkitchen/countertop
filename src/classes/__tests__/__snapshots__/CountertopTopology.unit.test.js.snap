// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountertopTopology #unit generateSourceStreams Should generate streams for any source stations 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
]
`;

exports[`CountertopTopology #unit generateStreams Should generate the correct streams 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
  Object {
    "mouth": 1,
    "source": 0,
    "tributaryMap": Map {
      "foo" => Object {
        "mouth": 0,
        "source": 0,
        "tributaryMap": Map {},
      },
    },
  },
]
`;

exports[`CountertopTopology #unit generateStreams Should not generate streams for stations with no inputs 1`] = `
Array [
  Object {
    "mouth": 0,
    "source": 0,
    "tributaryMap": Map {},
  },
]
`;
